File: Leetcode/python/sol112.py
# //https://leetcode.com/problems/path-sum/description/
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:

        if not root:
            return False

        Q = [(root, root.val)]

        while Q:
            node, suma = Q.pop()
            if node.left is None and node.right is None and suma == targetSum:
                return True

            if node.left:
                Q.append((node.left, suma+node.left.val))
            if node.right:
                Q.append((node.right, suma+node.right.val))


        return False


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        if not root:
            return False
        targetSum -= root.val
        if not root.left and not root.right and targetSum == 0:
            return True

        return self.hasPathSum(root.left, targetSum) or self.hasPathSum(root.right, targetSum)

File: Leetcode/javascript/sol112.js

//https://leetcode.com/problems/path-sum/


/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {boolean}
 */
var hasPathSum = function(root, targetSum) {

   function helper(node, target) {
        if (!node){
            return false;
        }
        if(!node.left && !node.right){
            return target == node.val;
        }
        let new_target = target - node.val
        return (helper(node.left, new_target) || helper(node.right, new_target))
    };

    return helper(root, targetSum);
};